man/arcfs.5
.TH ARCFS 5
.SH NAME
ArcFS \- Archive file format for the RISC OS ArcFS filesystem
.SH CONVENTIONS
Strings are given in ASCII with C-style escapes. No terminating \0 is
present if not explicitly included.

Words are 32 bit little-endian numbers.

Half words are 16 bit little-endian numbers.

This document describes format version 0, for ArcFS minimum versions 0.40
and 1.00, and version 0a, for ArcFS minimum versions 0.58 and 2.17.

.SH FORMAT
The file consists of a file header, a sequence of archive headers, then
a sequence of compressed files.

.SH FILE HEADER
.I File string signature
.br
The string "Archive" followed by a zero byte.

.I Header length
.br
Length of header info in bytes, one word, multiple of 36.

.I Data offset
.br
Offset, in bytes, into archive at which data starts, one word.

.I Read-only version
.br
Minimum read-only version required, multiplied by 100, one word.

.I Read-write version
.br
Minimum read--write version required, multiplied by 100, one word.

.I Reserved
.br
Reserved for future use, 68 bytes.

.SH ARCHIVE HEADER
The last header starts at 60+length of headers and ends at 96+length of
headers. The format of each header is as follows.

.I Info byte
.br
Object information byte, show compression method used:
.nf
	0x00	end of directory marker
	0x01	object has been deleted
	0x82	stored
	0x83	packed
	0x88	crunched
	0xFF	compressed
.fi

.I File name
.br
File name, 11 bytes.

.I Full length
.br
Original file length, or 0xFFFFFFFF for a directory), one word.

.I Load address
.br
Load address of the file, or 0xFFFFFFdd for a directory), one word.

.I Exec address
.br
Exec address of the file, or 0xdddddddd for a directory), one word.

File load and execution address may contain a filetype and datestamp as
normal.

Directories can be date-stamped, the number of centiseconds since the 1st
January 1900 when the directory was created are given by &dddddddddd
contained in the directory's load and execution addresses as shown above.

Archives converted from spark(5) format using ArcConv will have the same
full length, compressed length, load address and execute address as they
had in the Spark archive.

.I Attributes
.br
Attributes of the file, one word.
Contains the number of bits compression for crunch and compress
format files and the least significant bytes contains the standard
RISC OS information on file access.

.I Compressed length
.br
Length of compressed data, or 0xFFFFFFFF for a directory, one word.

.I Information word
.br
The top bit (definitively) indicates whether the object is a file or
a directory, 0 for a file, 1 for a directory.

In the case of a file, the least significant 31 bits show the offset from
the start of data at which the file data can be found.

In the case of a directory, the least significant 31 bits show the offset
from the start of headers (96) at which the next object in this directory
can be found, the following object being the first in the sub-directory.

.SH FORMAT VERSION 0a
This archive format is backwards compatible with versions 0.40 and 1.00, the
only difference being that there is a block of catalogue information
immediately prior to the start of each file's data as described below.

Note that this information is there to aid the recovery of most files in the
event of archive corruption and is not used by ArcFS.

Note, also that Compact and Autocompact in release versions 1.09, 2.02 and
2.11 will compact out the catalogue information blocks, so try to avoid
writing to new archives with old versions of ArcFS.

.nf
.I Marker
A marker to indicate the start of a file, the string "FileData", not terminated.

.I Info byte
.br
One byte, as above.

.I Leaf name
.br
The file name, 11 bytes.

.I Full length
.br
One word, as above.

.I Load address
.br
One word, as above.

.I Exec address
.br
One word, as above.

.I Attributes
.br
One word, as above.

.SH SEE ALSO
spark(5), nspark(1)


CHANGES

1.2.4	7th Oct 92	1st public release.

1.2.5	9th Oct 92	Fixed bug in misc.c::riscos_path() that caused nspark
			to core dump under SCO UNIX.

1.2.6	19th Oct 92	Added -x option as an alternative to -u.  Changed
			unix.c::filestamp() so that timelocal(3) is only used
			for SunOS, and mktime(3) is used for all other
			flavours of UNIX.

1.2.7	23rd Oct 92	Minor cosmetic changes... I should really leave this
			thing alone.

1.2.8	6th Nov 92	Fixed function definition for unarc.c::prompt_user()
			(was defined the ANSI C way).

			Fixed io.c::read_header() so it doesn't read the
			load/exec/attr fields if the archive isn't an
			Archimedes archive.
			
			Corrected comments in compress.c about what crunch
			and squash methods were.
			
			Slight change to algorithm in crc.c::calccrc(), though
			no functional changes made.

1.2.9	9th Nov 92	compress.c: Moved offset and size from getcode() into
			global scope so that uncompress() can re-initialise
			them each time it is called.

			pack.c/pack.h: Added putc_init() to initialise the
			static variables used by putc_ncr().

1.2.10	12th Nov 92	Fixed bug with realloc() size in misc.c: uplevel().
			(reported by arb@comp.lancs.ac.uk).

1.2.11	7th Dec 92	Reformatted all source files using Emacs C-Mode (4-
			space tabs etc.).  Slight change to unarc.c::unarc()
			logic when printing file details during unarchiving.

1.3Beta 8th Dec 92	Added -a option to allow filetypes to be appended
			to filenames as they are unarchived.  This is to allow
			Acorn <-> UNIX NFS to work correctly.
			(requested by dhmyrdal@pvv.unit.no)

1.3	9th Dec 92	Changed -a option to -T, made filetypes append in
			lowercase hex rather than uppercase.

1.3.1	9th Dec 92	Changed ret from int in io.c::check_stream().  #ifdef'd
			out write_halfword() and write_word() in io.c/io.h .

1.3.2	22nd Dec 92	misc.c: Changed #include <malloc.h> to
			#include <stdlib.h>

1.4	23rd Dec 92	unarc.c: added total-printing if in verbose mode.

NEW MAINTAINER: Andrew Brooks, arb@comp.lancs.ac.uk.

1.5	5th Mar 93	Added code to compile on RISC OS: mkfiles/acorn,
			acorn.c, changes to io.c, unarc.c, main.c, io.h to
			add prototypes needed.

1.6	20th Aug 93     main.c: Add -C option to convert filenames to lowercase
			io.c: allow top-bit-set characters in RISCOS filenames.
			compress.c, unarc.c: prevent output if in quiet mode.

1.7	20th Aug 93	Added code to read headers from ArcFS archives.
			Changes to io.c, unarc.c, mkfiles.  Added arcfs.[ch]
			WARNING: Code is unfinished: only the list option (-l)
			can be used (and timestamps will be incorrect).  Trying
			to test or extract from the archive will fail, and
			probably dump core.  Please add the necessary
			code if you can, I don't have time at the moment.

1.7.1   4th Mar 94	Fixed many bugs in ArcFS code: changes to unarc.c,
			compress.c also.  Still no timestamps, and files
			extracted are too long by a few null bytes for some
			reason (complen is too large, but why?), so CRC fails.

1.7.2   26th Oct 94	Fixed date and time conversion for ArcFS archives.
			Added manual pages supplied by Duncan (?)
			(m91dps@ecs.oxford.ac.uk).

1.7.2a	9th Nov 94	Altered unix makefile to call ./mkendian not mkendian
			as suggested by Tom Tromey (tromey@busco.lanl.gov).
			Added rename scripts for RISC OS/UNIX.

1.7.3	12th Dec 94	Finally fix ArcFS archive extraction, fixes in
			arcfs.c, unarc.c, io.c and crc.c.  Modifications
			to manual pages to correct errors and add new details.

1.7.4	6th Jan 95	Added Windows NT and Alpha support by Alun Jones
			(auj@aber.ac.uk).
			Restructured directories and added executables for
			RISC OS, Sun OS 4.1.3 and Solaris 2.3.
			Added executables for Linux, MIPS and Alpha supplied
			by Alun Jones.
			Updated manual pages for new file formats.
			Added text versions of manual pages for those without
			nroff.

1.7.5	25th Jan 95	Bug fixes in compress.c, thanks to all who reported
			the problem.  Included binaries that report version
			1.7.4 should not be exhibit the bug as it was caused
			by a last-minute source change (mae culpa!).

1.7.6   01th Aug 95     DOS port. (Beta by Bob Brand, bob@wop.wtb.tue.nl,
			or B.A.Brand@wtb.tue.nl, addresses valid until
			at least Oct 1st, 1996), fido: 2:2801/208.2.
                        Removed floating point operation in
			arcfs.c::arcfs_read_header().
			Made htab[] and codetab[] arrays in compress.c
			dynamically allocated arrays (DOS version only!).
			Made misc.c::basename() insensitive for null-
			pointers. (IRIX's strlen() and a few other string-
			functions coredump on null-pointers. Should not
			happen under normal execution. But it can happen
			in the DOS version running under DOS 2.x.)
			Renamed io.h to sparkio.h (conflict with Borland
			C/C++ io.h).
			Removed `Does not support DOS' from BUGS file.
			Added `Does not time/date stamp directories
			under DOS or WINNT' to BUGS file.
			Included executables for SGI IRIX 4.x and 5.x.
			
1.7.7   17th March 1999 -I option Mark Usher, marku@magnet.at
		The -I option has been added to extract the individual
		files with an extra .inf file that contains the
		information relevant to that file. This should help
		people who use BBC emulators to read the files into
		disc images.
		The WinNT directory has been changed to Win32.



arcfs.c

/*
 * Handle ArcFS format archives
 *
 * Author: Andrew Brooks, arb@comp.lancs.ac.uk
 *
 * $Header: arcfs.c 1.5 95/08/01 $
 * $Log:        arcfs.c,v $
 * Revision 1.5  95/08/01  xx:xx:xx  BB
 * Fixes for Borland C/C++
 * Removed use of floating point routines.
 *
 * Revision 1.4  95/01/06  11:58:45  arb
 * Fixes for Alpha.
 *
 * Revision 1.3  94/12/12  17:25:25  arb
 * Fixes for writesize.
 *
 * Revision 1.2  94/10/26  15:06:35  arb
 * Fixed date and time conversion.
 *
 * Revision 1.1  94/02/28  21:41:23  arb
 * Fixed header, fixed include ordering, fixed directory check,
 * added seek to start of compressed data, fixed maxbits, ...
 * ie. got the thing working at last!
 *
 * Revision 1.0  93/08/20  12:40:15  arb
 * Initial revision
 */

#include <stdio.h>
#include <time.h>
#include "spark.h"
#ifdef UNIX
#include <sys/types.h>
#endif							/* UNIX */

/* BB changed next line */

/* #ifdef RISCOS */
#if defined(RISCOS) || defined(__MSDOS__)
#include <stdlib.h>
#include <string.h>
#endif	/* RISCOS */						/* || __MSDOS__ */
#include "arcfs.h"

/* BB changed next line because of conflict with Borland's io.h. */

/* #include "io.h" */
#include "nsparkio.h"

/* BB added next include for print_header(). */
#ifdef __MSDOS__
#include "misc.h"
#endif							/* __MSDOS__ */

#ifndef SEEK_SET
#define SEEK_SET 0
#endif

#ifdef UNIX
static char rcsid[] = "$Header: arcfs.c 1.5 95/08/01 $";
#endif							/* UNIX */


/*
 * Public flag to indicate whether the current archive is ArcFS format
 */
int arcfs = 0;


/*
 * Public number of compression bits, used in compress.c
 */
int arcfs_maxbits = 0;


/*
 * Public size of file being extracted, used in io.c, crc.c
 */
long writesize;
long crcsize;


/*
 * ArcFS header list element
 */
struct arcfs_header_s
{
	struct arcfs_header_s *next;
	Byte maxbits;
	Byte is_dir;
	Byte info_byte;
	Word info_word;
	Word seek;
	Header *header;
};
typedef struct arcfs_header_s *arcfs_header;


/*
 * Info byte
 */
#define AFS_ENDDIR   0
#define AFS_DELETED  1
#define AFS_STORE    0x82
#define AFS_PACK     0x83
#define AFS_CRUNCH   0x88
#define AFS_COMPRESS 0xFF


/*
 * Static data
 */
static int arcfs_initialised = 0;
static arcfs_header header_list = NULL;
static arcfs_header header_ptr = NULL;


/*
 * Convert RISC OS time to UNIX time.
 * RISC OS time is five bytes of centiseconds since 1900.
 * UNIX time is seconds since 1970.
 * MSB of RISC OS time is LSB of `load' plus `exec'.
 */

/* BB added extra prototype for Borland C/C++ */
#ifdef __MSDOS__
struct tm *rotm P__((Word load, Word exec));
#endif							/* __MSDOS__ */

struct tm *
rotm(load, exec)
	Word load, exec;
{
	Word low, high;
	time_t t;

	high = (load & 0xff) - 0x33l;
	low = exec - 0x6e996a00l;
	/* BB changed constant in next line to long */
	t = (time_t) (high * 42949673L + low / 100L);
	return (localtime(&t));
}


/*
 * Convert RISC OS time to SPARK time
 */

/* BB added extra prototype for Borland C/C++ */
#ifdef __MSDOS__
void arcfs_fixtime P__((Header * hdr));
#endif							/* __MSDOS__ */

void
arcfs_fixtime(hdr)
	Header *hdr;
{
	/* BB next line commented out, variable ti never used */
	/* time_t ti; */
	struct tm *tim;

	/* Convert to UNIX time first (as it is easy) */
	tim = rotm(hdr->load, hdr->exec);

	/* Convert UNIX time to SPARK time */
	hdr->date = (tim->tm_year - 80) << 9;
	hdr->date |= (tim->tm_mon + 1) << 5;
	hdr->date |= (tim->tm_mday);
	hdr->time = (tim->tm_hour) << 11;
	hdr->time |= (tim->tm_min) << 5;
	hdr->time |= (tim->tm_sec);
}


/*
 * Read ArcFS header
 */
Header *
arcfs_read_header(ifp)
	FILE *ifp;
{
	static Header null_header;
	static Word data_start;
	Word header_length = 0;
	Header *header;
	Word version;
	int i;
	Byte info_byte, name[12];
	Word length, load, exec, attr, complen, info_word;
	arcfs_header header_prev;
	int j;

	/* Return next header from list */
	if (arcfs_initialised)
	{
		/* If end of list return an empty header structure to indicate EOF */
		if (header_ptr == NULL)
			return (&null_header);

		/* Return next header in list */
		header = header_ptr->header;
		/* Seek to start of compressed data */
		if ((!header_ptr->is_dir)
			&& (fseek(ifp, (long) header_ptr->seek, SEEK_SET)))
		{
			printf("Cannot seek compressed data in this file\n");
			return (&null_header);
		}
		/* Set up number of compression bits */
		arcfs_maxbits = header_ptr->maxbits;
		/*if (header_ptr->is_dir) header = &null_header; */
		header_ptr = header_ptr->next;
		return (header);
	}

	/* Header list not constructed yet, so read all headers from file */
	arcfs_initialised = 1;
#ifdef BSD
	bzero((char *) &null_header, sizeof(null_header));
#else							/* not BSD */
	memset((char *) &null_header, '\0', sizeof(null_header));
#endif							/* BSD */
	null_header.comptype = 0;
	header_length = read_word(ifp);
	data_start = read_word(ifp);
	if ((version = read_word(ifp)) > 40)
	{
		/* BB removed floating point routines from next line
		   This saves linking the floating point routines under DOS
		   which yields quite a reduction in executable size.
		   And it removes the need to have the FPE present under RISC OS. */
		/* printf("Archive created by a newer version of ArcFS (%.2f)\n",((float)version)/100); */
		printf("Archive created by a newer version of ArcFS (%d.%02d)\n",
			   version / 100, version % 100);
		return (&null_header);
	}
	read_word(ifp);				/* read/write version */
	if ((version = read_word(ifp)) > 0)
	{
		printf("Archive format %d not understood\n", version);
		return (&null_header);
	}
	for (i = 0; i < 17; i++)
		read_word(ifp);			/* reserved */

	/* Read list of headers */
	for (i = 0; i < header_length / 36; i++)
	{
		/* Create list item */
		header = (Header *) malloc(sizeof(Header));
		header_ptr = (arcfs_header) malloc(sizeof(struct arcfs_header_s));
		if ((header == NULL) || (header_ptr == NULL))
		{
			printf("Out of memory\n");
			return (&null_header);
		}

		/* Read ArcFS file header */
		info_byte = read_byte(ifp);
		for (j = 0; j < 11; j++)
		{
			name[j] = read_byte(ifp);
			if (name[j] == PATHSEP)
				name[j] = '_';
			if (name[j] < ' ' || name[j] > '~')
				name[j] = '\0';
		}
		name[j] = '\0';
		length = read_word(ifp);
		load = read_word(ifp);
		exec = read_word(ifp);
		attr = read_word(ifp);
		complen = read_word(ifp);
		info_word = read_word(ifp);

		/* Examine, and create nspark header */
		if (info_byte == AFS_DELETED)
		{
			free(header);
			free(header_ptr);
			continue;
		}
		/* BB changed next line for Borland C/C++ 4 */
		/* header_ptr->is_dir = (info_word >> 31); */
#ifdef __MSDOS__
		header_ptr->is_dir = (Halfword) (info_word >> 31);
#else
		header_ptr->is_dir = (info_word >> 31);
#endif
		header_ptr->info_byte = info_byte;
		header_ptr->info_word = info_word;
		/* BB changed next line for Borland C/C++ 4 */
		/* header_ptr->maxbits = (attr & 0xff00) >> 8; */
#ifdef __MSDOS__
		header_ptr->maxbits = (Halfword) (attr & 0xff00) >> 8;
#else
		header_ptr->maxbits = (attr & 0xff00) >> 8;
#endif
		/* BB changed constant in next line to long. */
		header_ptr->seek = (info_word & 0x7fffffffL) + data_start;
		header->comptype = info_byte;
		strcpy(header->name, (char *) name);
		header->complen = complen;
		header->date = 0;
		header->time = 0;
		header->crc = (Halfword) (attr >> 16);
		header->origlen = length;
		header->load = load;
		header->exec = exec;
		header->attr = attr & 0xff;

		arcfs_fixtime(header);

		if (info_byte == AFS_ENDDIR)
		{
			/* Just return comptype == 0 */
			*header = null_header;
			header_ptr->is_dir = 0;
			header_ptr->seek = 0;
		}

		/* If it is an ArcFS directory then convert to a Spark directory */
		if (header_ptr->is_dir)
		{
			/* Make sure filetype is DDC */
			header->comptype = CT_NOTCOMP2;
			/* BB changed constant in next line to long. */
			header->load = 0xfffddcffL;
		}

		/* Add list item to list */
		/* Doing it here ensures that deleted items are not added */
		header_ptr->header = header;
		if (header_list == NULL)
			header_list = header_ptr;
		else
			header_prev->next = header_ptr;
		header_prev = header_ptr;
#ifdef DEBUGGING
		print_header(header);
#endif
	}

	/* Return first element */
	header_ptr = header_list;
	header = header_ptr->header;
	/* Seek to start of data for first element */
	if ((!header_ptr->is_dir)
		&& (fseek(ifp, (long) header_ptr->seek, SEEK_SET)))
	{
		printf("Cannot seek compressed data in this file\n");
		return (&null_header);
	}
	/* Set up number of compression bits */
	arcfs_maxbits = header_ptr->maxbits;
	/*if (header_ptr->is_dir) header = &null_header; */
	header_ptr = header_ptr->next;
	return (header);
}
